doc/api/diagnostics_channel.md

```diff
@@ -0,0 +1,189 @@
+# Diagnostics Channel
+
+<!--introduced_in=REPLACEME-->
+
+> Stability: 1 - Experimental
+
+<!-- source_link=lib/diagnostics_channel.js -->
+
+The `diagnostics_channel` module provides an API to report diagnostic data
+via named channels. It can be accessed using:
+
+```js
+const diagnostics_channel = require('diagnostics_channel');
+```
+
+## Public API
+
+### Overview
+
+Following is a simple overview of the public API.
+
+```js
+const diagnostics_channel = require('diagnostics_channel');
+
+// Subscribe to a channel by name
+diagnostics_channel.subscribe('my-channel', (name, data) => {
+  // Received data
+});
+
+// Subscribe to any channels matching a function result
+diagnostics_channel.subscribe({
+  shouldListen(name) {
+    return /^\s+-channel/.test(name);
+  },
+  onMessage(name, data) {
+    // Received data
+  }
+});
+
+// Subscribe to every channel
+diagnostics_channel.subscribe((name, data) => {
+  // Received data
+});
+
+// Get a reusable channel object
+const channel = diagnostics_channel.publisher('my-channel');
+
+// Subscribe directly to a channel object
+channel.subscribe((name, data) => {
+  // Received data
+});
+
+// Check if the channel has an active subscriber
+if (channel.shouldPublish()) {
+  // Publish data to the channel
+  channel.publish({
+    some: 'data'
+  });
+}
+
+// An alternative way to check for active subsribers before publishing
+channel.ifActive?.publish('my-channel', {
+  some: 'data'
+});
+```
+
+#### `diagnostics_channel.subscribe([name, ] onMessage)`
+
+* `name` {string} The channel name
+* `onMessage` {Function} The handler to receive channel messages
+
+Register a message handler to subscribe to existing channels and any future
+channels. The message handler can be a function or an object which includes
+an `onMessage` function. In object form, a `shouldListen` function may also
+be optionally included to filter which channels to subscribe to by running
+the function on the name of the channel whenever a new channel is created.
+
+```js
+const diagnostics_channel = require('diagnostics_channel');
+
+// Subscribe to a channel by name
+diagnostics_channel.subscribe('my-channel', (name, data) => {
+  // Received data
+});
+
+// Subscribe to any channels matching a function result
+diagnostics_channel.subscribe({
+  shouldListen(name) {
+    return /^\s+-channel/.test(name);
+  },
+  onMessage(name, data) {
+    // Received data
+  }
+});
+
+// Subscribe to every channel
+diagnostics_channel.subscribe((name, data) => {
+  // Received data
+});
+```
```

Comment:
Please add an `unsubscribe` equivalent, e.g. how can we avoid the memory leak.

Reply:
I can add an unsubscribe, if that's something we want. I just didn't add it because, at least for the use case _I_ see, I don't really see a case where one _would_ unsubscribe from a channel. APM vendors would just subscribe to all the channels the are interested in and leave the subscription active for as long as the app is running, so unsubscribing didn't seem important.

Reply:
I think this would be a good addition, the use case I see is sub/unsub to these at runtime if you only want to do so on demand.  Also might be useful of some testing scenarios and I am sure there are error conditions you would want to unsubscribe when you hit.